File: src/commands/clone.ts
import type { App } from '@/main/app.js'
import { exec, execAsync, hasFlag, logCommand } from '@/utils/cmd.js'
import { cwd, maxItems } from '@/utils/constants.js'
import { InvalidParamError, NotFoundError } from '@/utils/errors.js'
import { getLockfile } from '@/utils/lockfile.js'
import { verifyPromptResponse } from '@/utils/prompt.js'
import { existsSync } from 'node:fs'
import path from 'node:path'
import { detect as detectPackageManager } from '@antfu/ni'
import axios from 'axios'
import * as p from '@clack/prompts'

export interface Repository {
  name: string
  clone_url: string
  updated_at: string
  [k: string]: string
}

type PackageManager = 'npm' | 'yarn' | 'pnpm'

async function cloneCommand(params: string[], flags: string[]): Promise<void> {
  const repository = await getRepository(params, flags)

  const projectPath = await formatProjectPath(repository.name)
  const isCloned = existsSync(projectPath)

  if (isCloned) {
    logCommand(`cd ${projectPath}`)
    process.chdir(projectPath)
  } else {
    exec(`git clone ${repository.clone_url} ${projectPath}`)
    logCommand(`cd ${projectPath}`)
    process.chdir(projectPath)
    exec('git remote rename origin o')
  }

  const isNodeProject = existsSync(path.resolve(projectPath, 'package.json'))
  const isGoProject =
    !isNodeProject && existsSync(path.resolve(projectPath, 'go.mod'))

  if (isNodeProject) {
    const packageManager =
      (await detectPackageManager()) ?? (await selectPackageManagerPrompt())
    exec(`${packageManager} install`)
  } else if (isGoProject) {
    exec('go mod download')
  }

  exec('code .')
}

async function getRepository(
  params: string[],
  flags: string[],
): Promise<Repository> {
  const isGithubUrl = /github\.com.+\.git$/.test(params[0])

  if (isGithubUrl) {
    const repositoryUrl = params[0]

    return {
      clone_url: repositoryUrl,
      name: repositoryUrl.replace(/.+\/(.+)\.git$/, '$1'),
      updated_at: '',
    }
  }

  const repositories = await getUserRepositories()
  const isFilter = hasFlag(['--filter', '-f'], flags)

  if (params.length > 0 && isFilter) {
    const filterRegex = new RegExp(params.join('|'), 'i')
    const filteredRepositories = repositories.filter(repository =>
      filterRegex.test(repository.name),
    )

    if (filteredRepositories.length === 0) {
      throw new InvalidParamError('filter param', 'No repository found')
    }

    if (filteredRepositories.length === 1) {
      return filteredRepositories[0]
    }

    return selectRepositoryPrompt(filteredRepositories)
  }

  if (params.length > 0) {
    const repositoryName = params[0]

    const repositoryRegex = new RegExp(repositoryName, 'i')
    const filteredRepositories = repositories.filter(repository =>
      repositoryRegex.test(repository.name),
    )

    if (filteredRepositories.length === 0) {
      throw new NotFoundError(repositoryName)
    }

    const desiredRepository =
      filteredRepositories.find(
        repository => repository.name === repositoryName,
      ) ?? filteredRepositories[0]

    return desiredRepository
  }

  return selectRepositoryPrompt(repositories)
}

async function getUserRepositories(): Promise<Repository[]> {
  const getGithubCliRepositories = async (): Promise<Repository[]> => {
    const authStatus = await execAsync('gh auth status')

    if (!authStatus.includes('Logged in')) {
      return []
    }

    const repositoriesData = await execAsync('gh repo list -L 50')
    const repositories = repositoriesData
      .split('\n')
      .filter(Boolean)
      .map(line => {
        const username = line.replace(/^(.+?)\/.+/, '$1')
        const repositoryName = line.replace(/^.+?\/(.+?)\s.+/, '$1')
        const updatedAt = line.replace(/.+\s(.+)$/, '$1')

        return {
          name: repositoryName,
          clone_url: `https://github.com/${username}/${repositoryName}.git`,
          updated_at: updatedAt,
        } satisfies Repository
      })

    return repositories
  }

  const getGithubApiRepositories = async (): Promise<Repository[]> => {
    const username = await getLockfile('userGithubName')
    const { data: repositories } = await axios.get<Repository[]>(
      `https://api.github.com/users/${username}/repos`,
      { data: { username } },
    )

    return repositories
  }

  const githubRepositories = await Promise.all([
    getGithubApiRepositories(),
    getGithubCliRepositories(),
  ])

  const deduplicatedRepositories = githubRepositories
    .flat()
    .filter(
      (repo, i, arr) => i === arr.findIndex(_repo => _repo.name === repo.name),
    )

  return deduplicatedRepositories
}

async function formatProjectPath(repositoryName: string): Promise<string> {
  return hasFlag('--root')
    ? path.resolve(
        // eslint-disable-next-line unicorn/no-await-expression-member
        (await getLockfile('projectsRootList'))[0],
        repositoryName,
      )
    : path.resolve(cwd, repositoryName)
}

async function selectRepositoryPrompt(
  repositories: Repository[],
): Promise<Repository> {
  const sortedRepositories = repositories.sort((a, b) => {
    const date = [
      new Date(a.updated_at).getTime(),
      new Date(b.updated_at).getTime(),
    ]

    return date[0] > date[1]
      ? -1
      : date[0] < date[1]
        ? 1
        : a.name.localeCompare(b.name)
  })

  const response = await p.select({
    message: 'Select one of your repositories:',
    options: sortedRepositories.map(repository => ({
      label: repository.name,
      value: repository,
    })),
    initialValue: sortedRepositories[0],
    maxItems,
  })
  verifyPromptResponse(response)

  return response
}

async function selectPackageManagerPrompt(): Promise<PackageManager> {
  const options: PackageManager[] = ['pnpm', 'yarn', 'npm']
  const response = await p.select({
    message: 'Select your package manager:',
    options: options.map(pm => ({ value: pm })),
    initialValue: options[0],
    maxItems,
  })
  verifyPromptResponse(response)

  return response
}

export function cloneRecord(app: App): void {
  app.register({
    name: 'clone',
    alias: null,
    params: ['<repository>'],
    flags: ['--root', '--filter', '-f'],
    description:
      "Clone a Github's repository based on `setup`, sets git `origin` to `o`, install dependencies, and open it on vscode",
    example: 'my clone my-cli',
    action: cloneCommand,
  })
}

File: cmd/clone.go
/*
Copyright © 2024 NAME HERE <EMAIL ADDRESS>
*/
package cmd

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/Mist3rBru/go-clack/prompts"
	"github.com/Mist3rBru/my-cli/internals/lockfile"
	"github.com/Mist3rBru/my-cli/internals/utils"
	"github.com/Mist3rBru/my-cli/third_party/assert"
	"github.com/Mist3rBru/my-cli/third_party/ni"
	"github.com/spf13/cobra"
)

// cloneCmd represents the clone command
var cloneCmd = &cobra.Command{
	Use:   "clone",
	Short: "Clone a Github's repository",
	Long:  "Clone a Github's repository based on `setup`, sets git `origin` to `o`, install dependencies, and open it on vscode",
	Run: func(cmd *cobra.Command, args []string) {
		repo := getRepository(cmd, args)
		projectPath := formatProjectPath(cmd, repo)

		if utils.Exists(projectPath) {
			prompts.Step(fmt.Sprintf("cd %s", projectPath))
			os.Chdir(projectPath)
		} else {
			utils.ExecOrExit("git clone", repo.Clone_url, projectPath)
			prompts.Step(fmt.Sprintf("cd %s", projectPath))
			os.Chdir(projectPath)
			utils.ExecOrExit("git remote rename origin o")
		}

		if utils.Exists(filepath.Join(projectPath, "package.json")) {
			pkgManager, err := ni.Detect(ni.DetectOptions{})
			if err != nil {
				pkgManager = selectPackageManagerPrompt()
			}
			utils.ExecOrExit(string(pkgManager), "install")
		} else if utils.Exists(filepath.Join(projectPath, "go.mod")) {
			utils.ExecOrExit("go mod download")
		}

		utils.ExecOrExit("code", projectPath)
	},
}

type Repository struct {
	Name       string `json:"name"`
	Clone_url  string `json:"clone_url"`
	Updated_at string `json:"updated_at"`
}

func getRepository(cmd *cobra.Command, args []string) *Repository {
	githubRegex := regexp.MustCompile(`github.com\.com.+\.git$`)

	if len(args) > 0 && githubRegex.MatchString(args[0]) {
		repotitoryUrl := args[0]
		repositoryNameRegex := regexp.MustCompile(`.+\/(.+)\.git$`)
		repositoryName := repositoryNameRegex.ReplaceAllString(repotitoryUrl, "$1")

		return &Repository{
			Name:       repositoryName,
			Clone_url:  repotitoryUrl,
			Updated_at: "",
		}
	}

	repos := getUserRepositories()
	filter := cmd.Flag("filter").Value.String()

	if len(args) > 0 {
		for _, repo := range repos {
			if args[0] == repo.Name {
				return repo
			}
		}
		prompts.Warn("repository not found")
	} else if filter != "" {
		filterRegex := regexp.MustCompile("(?i)" + filter)
		var filteredRepos []*Repository
		for _, repo := range repos {
			if filterRegex.MatchString(repo.Name) {
				filteredRepos = append(filteredRepos, repo)
			}
		}
		assert.Assert(len(filteredRepos) > 0, "no filtered repository found")

		if len(filteredRepos) == 1 {
			return filteredRepos[0]
		}

		return selectRepositoryPrompt(filteredRepos)
	}

	return selectRepositoryPrompt(repos)
}

func getUserRepositories() []*Repository {
	var wg sync.WaitGroup
	var repos []*Repository
	reposCh := make(chan *Repository, 10)

	s := prompts.Spinner(context.Background(), prompts.SpinnerOptions{})
	s.Start("Fetching repositories")
	defer s.Stop("", 0)

	wg.Add(2)
	go getGhCliRepositories(&wg, reposCh)
	go getGhRepositories(&wg, reposCh)

	go func() {
		wg.Wait()
		close(reposCh)
	}()

	for repo := range reposCh {
		repos = append(repos, repo)
	}

	contains := func(list []*Repository, repo *Repository) bool {
		for _, item := range list {
			if item.Name == repo.Name {
				return true
			}
		}
		return false
	}

	removeDuplicates := func(repos []*Repository) []*Repository {
		list := []*Repository{}
		for _, repo := range repos {
			if !contains(list, repo) {
				list = append(list, repo)
			}
		}
		return list
	}

	return removeDuplicates(repos)
}

func getGhCliRepositories(wg *sync.WaitGroup, reposCh chan *Repository) {
	defer wg.Done()

	authStatus, err := utils.ExecSilent("gh auth status")
	if err != nil || !strings.Contains(authStatus, "Logged in") {
		return
	}

	output, err := utils.ExecSilent("gh repo list -L 50")
	if err != nil {
		return
	}

	for _, line := range strings.Split(output, "\n") {
		if line == "" {
			continue
		}
		userName := regexp.MustCompile(`^(.+?)\/.+`).ReplaceAllString(line, "$1")
		repositoryName := regexp.MustCompile(`^.+?\/(.+?)\s.+`).ReplaceAllString(line, "$1")
		updatedAt := regexp.MustCompile(`.+[\s\t](.+)$`).ReplaceAllString(line, "$1")
		// fmt.Println(userName, repositoryName, updatedAt)

		reposCh <- &Repository{
			Name:       repositoryName,
			Clone_url:  fmt.Sprintf("https://github.com/%s/%s.git", userName, repositoryName),
			Updated_at: updatedAt,
		}
	}
}

func getGhRepositories(wg *sync.WaitGroup, reposCh chan *Repository) {
	defer wg.Done()

	l := lockfile.Open()
	defer l.Close()

	userName := l.GetUserGithubName()
	if userName == "" {
		return
	}

	resp, err := http.Get(fmt.Sprintf("https://api.github.com/users/%s/repos", userName))
	if err != nil {
		return
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return
	}

	var repos []Repository
	err = json.Unmarshal(body, &repos)
	if err != nil {
		return
	}

	for _, repo := range repos {
		reposCh <- &repo
	}
}

func selectRepositoryPrompt(repos []*Repository) *Repository {
	sort.Slice(repos, func(i, j int) bool {
		a, err := time.Parse("2006-01-02T15:04:05Z0700", repos[i].Updated_at)
		assert.NoError(err, "invalid repository updated_at field")
		b, err := time.Parse("2006-01-02T15:04:05Z0700", repos[j].Updated_at)
		assert.NoError(err, "invalid repository updated_at field")
		return a.After(b)
	})

	var options []prompts.SelectOption[*Repository]
	for _, repo := range repos {
		options = append(options, prompts.SelectOption[*Repository]{
			Label: repo.Name,
			Value: repo,
		})
	}
	repo, err := prompts.Select[*Repository](prompts.SelectParams[*Repository]{
		Message: "Select one of your repositories:",
		Options: options,
	})
	utils.VerifyPromptCancel(err)

	return repo
}

func formatProjectPath(cmd *cobra.Command, repo *Repository) string {
	l := lockfile.Open()
	defer l.Close()

	if isRoot, _ := cmd.Flags().GetBool("root"); isRoot {
		projectsRootList := l.GetUserProjectsRootList()
		return filepath.Join(projectsRootList[0], repo.Name)
	}
	cwd, _ := os.Getwd()
	return filepath.Join(cwd, repo.Name)
}

func selectPackageManagerPrompt() ni.Agent {
	agent, err := prompts.Select[ni.Agent](prompts.SelectParams[ni.Agent]{
		Message: "Select your package manager:",
		Options: []prompts.SelectOption[ni.Agent]{
			{Label: "pnpm", Value: "pnpm"},
			{Label: "yarn", Value: "yarn"},
			{Label: "npm", Value: "npm"},
		},
	})
	utils.VerifyPromptCancel(err)
	return agent
}

func init() {
	rootCmd.AddCommand(cloneCmd)

	cloneCmd.Flags().StringP("filter", "f", "", "Filters the repository options with the given value")

	cloneCmd.Flags().BoolP("root", "r", false, "Change the target folder from cwd to the first project root stored on lockfile")

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	// cloneCmd.PersistentFlags().String("foo", "", "A help for foo")

	// Cobra supports local flags which will only run when this command
	// is called directly, e.g.:
	// cloneCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
}
